File: android/beaconsdk/src/main/java/com/xenber/beaconsdk/network/GatewayClient.kt
package com.xenber.beaconsdk.network

import android.content.ContentResolver
import android.provider.Settings
import com.xenber.beaconsdk.BeaconConfig
import com.xenber.beaconsdk.util.Logger
import com.xenber.beaconsdk.diagnostic.SdkTracker
import com.xenber.beaconsdk.diagnostic.SdkErrorCodes
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import kotlin.concurrent.thread

class GatewayClient(
    private val config: BeaconConfig,
    private val contentResolver: ContentResolver
) {

    private val sdkTracker = SdkTracker()

    interface GatewayCallback {
        fun onSuccess(response: JSONObject)
        fun onError(code: Int?, message: String)
    }

    fun fetchBeacons(callback: (Map<String, String>) -> Unit) {
        val dataUrl = config.dataUrl?.trim() ?: return

        if (!dataUrl.startsWith("http")) {
            Logger.e("Invalid data URL")
            return
        }

        thread {
            try {
                val conn = URL(dataUrl).openConnection() as HttpURLConnection
                conn.apply {
                    requestMethod = "GET"
                    setRequestProperty("Accept", "application/json")
                    // config.authToken?.takeIf { it.isNotBlank() }?.let {
                    //     setRequestProperty("Authorization", "Bearer $it")
                    // }
                    connectTimeout = 10_000
                    readTimeout = 10_000
                }

                val code = conn.responseCode
                if (code == 200) {
                    val responseText = conn.inputStream.bufferedReader().readText()
                    val json = JSONObject(responseText)
                    val dataArray = json.getJSONArray("data")

                    val beaconMap = mutableMapOf<String, String>()
                    for (i in 0 until dataArray.length()) {
                        val beacon = dataArray.getJSONObject(i)
                        val mac = beacon.getString("beacon_mac").uppercase()
                        val name = beacon.getString("location_name")
                        beaconMap[mac] = name
                    }
                    Logger.i("Fetched ${beaconMap.size} beacons from server")
                    callback(beaconMap)
                } else {
                    Logger.e("Failed to fetch beacons, code: $code")
                }
            } catch (e: Exception) {
                SdkTracker().error(
                    SdkErrorCodes.NETWORK_ERROR,
                    "Data fetch error: ${e.message}"
                )
                throw e
                Logger.e("Network error during beacon sync: ${e.message}")
            }
        }
    }


    fun sendDetection(
        mac: String,
        rssi: Int,
        battery: Int?,
        isInitial: Boolean = true,
        callback: GatewayCallback
    ) {
        val body = JSONObject().apply {
            put("type", "beacon_detection")
            put("user_id", config.userId)
            put("phone_id", getDeviceId())
            put("beacon_mac", mac)
            put("rssi", rssi)
            put("battery", battery ?: JSONObject.NULL)
            put("is_initial", isInitial)
            put("timestamp", getKLTimestamp())
        }

        executeRequest(body, callback)
    }

    fun sendEvent(
        eventType: String,
        details: JSONObject,
        callback: GatewayCallback? = null
    ) {
        val body = JSONObject().apply {
            put("type", "gateway_event")
            put("event_type", eventType)
            put("user_id", config.userId)
            put("phone_id", getDeviceId())
            put("details", details)
            put("timestamp", getKLTimestamp())
        }

        executeRequest(body, callback)
    }

    private fun executeRequest(
        body: JSONObject,
        callback: GatewayCallback?
    ) {
        val gatewayUrl = config.gatewayUrl.trim()
        if (!gatewayUrl.startsWith("http")) {
            callback?.onError(null, "Invalid gateway URL")
            return
        }

        thread {
            try {
                val conn = URL(gatewayUrl).openConnection() as HttpURLConnection
                val payload = body.toString().toByteArray(Charsets.UTF_8)

                Logger.i("📤 [Gateway] Sending ${body.optString("type")}: $body")

                conn.apply {
                    requestMethod = "POST"
                    setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                    setRequestProperty("Accept", "application/json")
                    // config.authToken?.takeIf { it.isNotBlank() }?.let {
                    //     setRequestProperty("Authorization", "Bearer $it")
                    // }

                    setRequestProperty("Content-Length", payload.size.toString())
                    doOutput = true
                    doInput = true
                    useCaches = false
                    connectTimeout = 10_000
                    readTimeout = 10_000
                }

                conn.outputStream.use { it.write(payload) }

                val code = conn.responseCode
                val responseText = try {
                    if (code in 200..299)
                        conn.inputStream.bufferedReader().readText()
                    else
                        conn.errorStream?.bufferedReader()?.readText()
                } catch (e: Exception) {
                    null
                }

                if (code == 200 && responseText != null) {
                    callback?.onSuccess(JSONObject(responseText))
                } else {
                    callback?.onError(code, responseText ?: "Empty response")
                }

            } catch (e: Exception) {
                SdkTracker().error(
                    SdkErrorCodes.NETWORK_ERROR,
                    "Gateway transport error: ${e.message}"
                )
                throw e
                Logger.e("Gateway transport error: ${e.message}")
                callback?.onError(null, e.message ?: "Gateway error")
            }   
        }
    }

    private fun getDeviceId(): String {
        return Settings.Secure.getString(contentResolver, Settings.Secure.ANDROID_ID) ?: "unknown_device"
    }

    private fun getKLTimestamp(): String {
        return ZonedDateTime
            .now(ZoneId.of("Asia/Kuala_Lumpur"))
            .format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
    }
}

File: ios/Runner/BeaconSdk/Monitoring/GatewayClient.swift
import Foundation
import UIKit

class GatewayClient {

  private var config: BeaconConfig?
  private let session: URLSession
  private let sdkTracker = SdkTracker()
  private let prefs = PreferenceStore()
  
  typealias JSONCallback = ([String: Any]) -> Void
  typealias BeaconsCallback = ([String: String]) -> Void
  typealias GatewayCallback = (Result<[String: Any], Error>) -> Void

  init() {
    let sessionConfig = URLSessionConfiguration.default
    sessionConfig.timeoutIntervalForRequest = 10.0
    sessionConfig.timeoutIntervalForResource = 10.0
    self.session = URLSession(configuration: sessionConfig)
  }

  func configure(_ config: BeaconConfig) {
    self.config = config
  }

  func fetchBeacons(completion: @escaping BeaconsCallback) {
    guard let dataUrlString = config?.dataUrl.trimmingCharacters(in: .whitespacesAndNewlines),
          let url = URL(string: dataUrlString),
          dataUrlString.hasPrefix("http") else {
      Logger.e("Invalid data URL")
      completion(prefs.getTargetBeacons())
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setValue("application/json", forHTTPHeaderField: "Accept")

    let task = session.dataTask(with: request) { [weak self] data, response, error in
      guard let self = self else { return }

      if let error = error {
        self.handleNetworkError(error, context: "Fetch Beacons")
        completion(self.prefs.getTargetBeacons())
        return
      }

      guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200, let data = data else {
        Logger.e("Failed to fetch beacons, status: \((response as? HTTPURLResponse)?.statusCode ?? 0)")
        completion(self.prefs.getTargetBeacons())
        return
      }

      do {
        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let dataArray = json["data"] as? [[String: Any]] {
          
          var beaconMap: [String: String] = [:]
          
          for item in dataArray {
            if let uuid = item["beacon_uuid"] as? String,
               let name = item["location_name"] as? String {
                let major = item["beacon_major"] as? Int
                let minor = item["beacon_minor"] as? Int

                var key = uuid.uppercased()

                if let maj = major, let min = minor {
                  key = "\(uuid.uppercased()):\(maj):\(min)"
                  Logger.d("📡 Mapped: \(key) -> \(name)")
                } else {
                  Logger.d("📡 Mapped: \(key) (Region Only) -> \(name)")
                }

                beaconMap[key] = name
              }
          }
            
          Logger.i("Fetched \(beaconMap.count) beacons from server")
          self.prefs.saveTargets(beaconMap)
          completion(beaconMap)
        } else {
          ompletion(self.prefs.getTargetBeacons())
        }
      } catch {
        self.handleNetworkError(error, context: "JSON Parse")
        completion(self.prefs.getTargetBeacons())
      }
    }
    task.resume()
  }

  // func sendDetection(mac: String, rssi: Int, timestamp: Int, battery: Int? = nil, isInitial: Bool = true, completion: @escaping JSONCallback) {
  //     guard let config = config else { return }
      
  //     let components = mac.components(separatedBy: ":")
  //     let uuid = components[0]
  //     let major = components.count > 1 ? Int(components[1]) : nil
  //     let minor = components.count > 2 ? Int(components[2]) : nil

  //     let body: [String: Any] = [
  //       "type": "beacon_detection",
  //       "user_id": config.userId,
  //       "phone_id": getDeviceId(),

  //       // Send separated fields 
  //       "beacon_uuid": uuid,
  //       "major": major ?? NSNull(),
  //       "minor": minor ?? NSNull(),

  //       "beacon_mac": mac,
  //       "rssi": rssi,
  //       "battery": battery ?? NSNull(), 
  //       "is_initial": isInitial,
  //       "timestamp": getKLTimestamp()  
  //     ]
      
  //     executeRequest(body: body) { result in
  //     switch result {
  //     case .success(let json):
  //       completion(json)
  //     case .failure(_):
  //       completion([:])
  //     }
  //   }
  // }

  // func sendEvent(_ eventType: String, data: [String: Any]) {
  //   guard let config = config else { return }
    
  //   let body: [String: Any] = [
  //     "type": "gateway_event",
  //     "event_type": eventType,
  //     "user_id": config.userId,
  //     "phone_id": getDeviceId(),
  //     "details": data,
  //     "timestamp": getKLTimestamp()
  //   ]
    
  //   executeRequest(body: body) { _ in }
  // }

  // private func executeRequest(body: [String: Any], completion: @escaping (Result<[String: Any], Error>) -> Void) {
  //   guard let gatewayUrlString = config?.gatewayUrl.trimmingCharacters(in: .whitespacesAndNewlines),
  //         let url = URL(string: gatewayUrlString),
  //         gatewayUrlString.hasPrefix("http") else {
  //     completion(.failure(NSError(domain: "Gateway", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])))
  //     return
  //   }

  //   var request = URLRequest(url: url)
  //   request.httpMethod = "POST"
  //   request.setValue("application/json; charset=UTF-8", forHTTPHeaderField: "Content-Type")
  //   request.setValue("application/json", forHTTPHeaderField: "Accept")
    
  //   do {
  //     request.httpBody = try JSONSerialization.data(withJSONObject: body, options: [])
  //     Logger.i("📤 [Gateway] Sending \(body["type"] ?? "unknown"): \(body)")
  //   } catch {
  //     Logger.e("Failed to serialize JSON")
  //     return
  //   }

  //   let task = session.dataTask(with: request) { [weak self] data, response, error in
  //     guard let self = self else { return }

  //     if let error = error {
  //       self.handleNetworkError(error, context: "Gateway Transport")
  //       completion(.failure(error))
  //       return
  //     }

  //     guard let httpResponse = response as? HTTPURLResponse else { return }

  //     var responseJSON: [String: Any] = [:]
  //     if let data = data {
  //       responseJSON = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
  //     }

  //     if (200...299).contains(httpResponse.statusCode) {
  //       completion(.success(responseJSON))
  //     } else {
  //       let msg = "HTTP \(httpResponse.statusCode)"
  //       Logger.e("Gateway error: \(msg)")
  //       completion(.failure(NSError(domain: "Gateway", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: msg])))
  //     }
  //   }
  //   task.resume()
  // }

  private func handleNetworkError(_ error: Error, context: String) {
    sdkTracker.error(.networkError, "\(context) error: \(error.localizedDescription)", error)
    Logger.e("\(context) error: \(error.localizedDescription)")
  }

  private func getDeviceId() -> String {
    return UIDevice.current.identifierForVendor?.uuidString ?? "unknown_device"
  }

  private func getKLTimestamp() -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
    formatter.timeZone = TimeZone(identifier: "Asia/Kuala_Lumpur")
    return formatter.string(from: Date())
  }
}